tests/test_fv3run.py

```diff
@@ -0,0 +1,243 @@
+import unittest
+import os
+import shutil
+import tempfile
+import contextlib
+import collections
+import subprocess
+from subprocess import check_call as original_check_call
+import pytest
+import yaml
+import fv3config
+
+TEST_DIR = os.path.dirname(os.path.realpath(__file__))
+MOCK_RUNSCRIPT = os.path.abspath(os.path.join(TEST_DIR, 'testdata/mock_runscript.py'))
+DOCKER_IMAGE_NAME = 'us.gcr.io/vcm-ml/fv3gfs-python'
+USER_UID_GID = f'{os.getuid()}:{os.getgid()}'
+
+MockTempfile = collections.namedtuple('MockTempfile', ['name'])
+
+MPI_ENABLED = (shutil.which('mpirun') is not None)
+try:
+
+    subprocess.check_call(
+        ['docker', 'image', 'inspect', DOCKER_IMAGE_NAME],
+        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
+    )
+    DOCKER_ENABLED = True
+except (subprocess.CalledProcessError, FileNotFoundError):
+    DOCKER_ENABLED = False
+
+
+def subprocess_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        subprocess.check_call(
+            ['fv3run', config_file.name, outdir, '--runfile', MOCK_RUNSCRIPT]
+        )
+
+
+def docker_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        try:
+            subprocess.check_call(
+                ['fv3run', config_file.name, outdir, '--runfile', MOCK_RUNSCRIPT,
+                 '--dockerimage', DOCKER_IMAGE_NAME]
+            )
+        except subprocess.CalledProcessError as err:
+            if os.path.isfile(os.path.join(outdir, 'stdout.log')):
+                with open(os.path.join(outdir, 'stdout.log')) as stdout:
+                    print(f'STDOUT: {stdout.read()}')
+            if os.path.isfile(os.path.join(outdir, 'stderr.log')):
+                with open(os.path.join(outdir, 'stderr.log')) as stderr:
+                    print(f'STDERR: {stderr.read()}')
+            raise err
+
+
+def config_dict_module_run(config_dict, outdir):
+    fv3config.run(config_dict, outdir, runfile=MOCK_RUNSCRIPT)
+
+
+def config_dict_filename_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        fv3config.run(config_file.name, outdir, runfile=MOCK_RUNSCRIPT)
+
+
+def count_executed_ranks(rundir):
+    filenames = os.listdir(rundir)
+    total = 0
+    while f'rank_{total}' in filenames:
+        total += 1
+    return total
+
+
+@contextlib.contextmanager
+def mocked_check_call():
```

Comment:
It seems a little complicated to write a context manager to mock out dependences. Both `unittest` and `pytest` have recommend methods for doing this. See https://docs.pytest.org/en/latest/monkeypatch.html or `unittest.patch`. Would that work instead of this context manager?

Reply:
What if we are using a platform that doesn't use an explicit call to mpirun?  The Crays will use srun to invoke MPI execution.  Do we expect to use this code on the supercomputers?

Reply:
Should we add MPI version checks here?  As there are ABI compatibility issues with certain versions.

Reply:
no problem!